zilliqa/src/pool.rs

```diff
@@ -0,0 +1,213 @@
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ReadyItem {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.gas_price.cmp(&other.gas_price)
+    }
+}
+
+impl From<&VerifiedTransaction> for ReadyItem {
+    fn from(txn: &VerifiedTransaction) -> Self {
+        ReadyItem {
+            gas_price: txn.tx.gas_price(),
+            from_addr: txn.signer,
+            nonce: txn.tx.nonce(),
+        }
+    }
+}
+
+impl TransactionPool {
+    /// Pop a *ready* transaction out of the pool, maximising the gas price.
+    ///
+    /// Ready means that the transaction has a nonce equal to the sender's current nonce or it has a nonce that is
+    /// consecutive with a previously returned transaction, from the same sender.
+    pub fn best_transaction(&mut self) -> Option<VerifiedTransaction> {
+        loop {
+            let Some(ReadyItem {
+                from_addr, nonce, ..
+            }) = self.ready.pop()
+            else {
+                return None;
+            };
+            let Some(transaction) = self.transactions.remove(&(from_addr, nonce)) else {
+                // A transaction might have been ready, but then we learnt that the sender's nonce increased, making it
+                // invalid. In this case, we will have removed the transaction from `transactions` in `update_nonce`.
+                // We loop until we find a transaction that hasn't been made invalid.
+                continue;
+            };
+            self.hash_to_index.remove(&transaction.hash);
+
+            if let Some(next_txn) = self.transactions.get(&(from_addr, nonce + 1)) {
+                self.ready.push(next_txn.into());
+            }
+
+            return Some(transaction);
+        }
+    }
+
+    pub fn insert_transaction(&mut self, transaction: VerifiedTransaction, account_nonce: u64) {
+        if transaction.tx.nonce() < account_nonce {
+            // This transaction is permanently invalid, so there is nothing to do.
+            return;
+        }
+
+        if transaction.tx.nonce() == account_nonce {
+            self.ready.push((&transaction).into());
+        }
+
+        self.hash_to_index.insert(
+            transaction.hash,
+            (transaction.signer, transaction.tx.nonce()),
+        );
+        self.transactions
+            .insert((transaction.signer, transaction.tx.nonce()), transaction);
+    }
+
+    pub fn get_transaction(&self, hash: Hash) -> Option<&VerifiedTransaction> {
+        let Some((addr, nonce)) = self.hash_to_index.get(&hash) else {
+            return None;
+        };
+
+        self.transactions.get(&(*addr, *nonce))
+    }
+
+    pub fn update_nonce(&mut self, address: Address, nonce: u64) {
+        // Remove a transaction from this sender with the previous nonce, if it exists.
+        self.transactions.remove(&(address, nonce - 1));
+
+        if let Some(next_txn) = self.transactions.get(&(address, nonce)) {
```

Comment:
this seems potentially bugprone because it assumes ready doesn't already have the (address, nonce) pairing somewhere in the heap, but it is possible.

Example: 
-Node 0 proposes a block, removing TX with nonce 0 and pushing TX with nonce 1 onto the heap (update nonce is called but I think doesn't really do anything as best_tx updated the state anyway).
-The state reverts to as it was before the block proposal.
-Normal TX gossip provides the same TX with nonce 0, this is now valid (race condition), and it goes somewhere into the heap.
-The heap now contains conflicting (address, _) 
-Later on, the mempool will incorrectly return TX with nonce 0 to execute.

I do note there is a check in consensus.rs that drops TXs with a nonce that is `<`... but presumably in the case your proposal fails to get accepted then you can get the TX with nonce 1, right?

Reply:
Reverts are now handled by re-creating the pool, which I think addresses this concern?

Reply:
Yes, seems like a reasonable way to handle it considering we don't care too much about the cost of forks